[STM]Add RNG (random number generator) for STM32F4 STM32F7 families

adustm opened the pull request:
These functions are to be tested in mbed-os environment.

The test was passed for DISCO_F429ZI and NUCLEO_F746ZG with IAR toolchain.
The compilation passes for DISCO_F429ZI / DISCO_F469NI / DISCO_F746NG / NUCLEO_F410RB / NUCLEO_F429ZI / NUCLEO_F746ZG / NUCLEO_F767ZI

**Please note that not every STM32F4 platforms are supporting the RNG feature. Some defines have been added in hal/targets/hal/TARGET_STM/TARGET_STM32F4/entropy_hardware_poll.c to enable the relevant targets.**
I did not find any code to test the function, thus I used the one below : 
```
#include "test_env.h"
#include "entropy_poll.h"

#define BUFF_SIZE       500
unsigned char myrndbuff[BUFF_SIZE]={0};
size_t *olen;

int main() {
	int ret = 0;
	int result = false;
	size_t mysize=0;
	olen = &mysize;
    ret = mbedtls_hardware_poll( NULL, myrndbuff, BUFF_SIZE, olen );
    if (ret == 0 ) {
    	printf("tls_ret_ok\n");
    	printf("%i\n",(int)*olen);
		if (*olen == BUFF_SIZE) {
			printf("size ok\n");
			result = true;
		} else printf("size ko\n");
    } else printf("tls_ret_ko\n");
}

```
This result was the following : 
```
tls_ret_ok
500
size ok
```
Another printf inside the code showed that the values seemed to be random (I trust the hardware :) )

adustm commented:
@screamerbg @pjbakker 
Could you please review and let me know what you think of it ?

sg- commented on hal/targets/hal/TARGET_STM/TARGET_STM32F4/entropy_hardware_poll.c:
```diff
@@ -0,0 +1,85 @@
+    /* Get Random byte */
+    for( i = 0; i < len; i++ ){
+        rng_get_byte( output + i );
+        printf("output %i: %i\n",i,(int)*(output+i));
```
left over debug info?

sg- commented:
formatting is a bit funky (see: https://developer.mbed.org/teams/SDK-Development/wiki/mbed-sdk-coding-style) and updated docs coming :)

adustm commented:
Hello Sam,
Thank you for the review.
Sorry for the coding style and docs (did you mean the doxygen comments ?), I copied it from the entropy done on the K64F  platform ;)
Is it better now ?

sg- commented:
/morph test

bridadan commented:
Looks like build failures for ARM toolchain for the following targets: NUCLEO_F429ZI	, DISCO_F469NI, DISCO_F429ZI, NUCLEO_F410RB:

```
Building project test_setup_failure (NUCLEO_F410RB, ARM)
Scan: ARM
Scan: test_setup_failure
Compile: main.cpp
Link: test_setup_failure
Error: L6218E: Undefined symbol mbedtls_hardware_poll (referred from entropy.o).
Finished: 0 information, 0 warning and 1 error messages.
```

0xc0170 commented:
@adustm Please can you check those failures?

adustm commented:
Hello everyone,
I just rebased the branch, merged the targets.json manually, and relaunched the build for  the following list of targets : DISCO_F429ZI / DISCO_F469NI / DISCO_F746NG / NUCLEO_F410RB / NUCLEO_F429ZI / NUCLEO_F746ZG / NUCLEO_F767ZI
You are right, there was an issue with the labels for STM32F4 family. I have fixed those labels in the targets.json file.
I looks like those defines (for instance STM32F429xx) where present in my previous environment, and not anymore... 

I am using the command : mbed test --compile -t ARM -m NUCLEO_F410RB and it compiles perfectly.

Let me know.
Kind regards
Armelle

bridadan commented:
Thanks @adustm!

/morph test

adustm commented:
Hello, Could you give me the command + result of the morph test, please ? I can't read them

adustm commented:
@0xc0170 , @bridadan 
is it possible to get details on the failure, please ?

bridadan commented:
Hi @adustm, sorry for the delay! There was a failure in the CI that occurred during this test, I will restart it.

/morph test

adustm commented:
Hello Brian, what is the failure now ? Could I have the command line + text of the result, so that I can test it, please ?
Kind regards
Armelle

screamerbg commented:
/morph test

0xc0170 commented:
The build 710 is without any failure

screamerbg commented:
@0xc0170 any idea why it reports failure?

bridadan commented:
Hi everyone, sorry for the delay on reporting back on this.

For build 710 (that last one), a test box died again which cause the whole thing to fail. I've since reorganized the boards so they are more balanced across the boxes, so that should help with stability.

I can confirm that build 685 looks ok, so the merge to master should be fine :+1:

adustm commented:
Hello, thanks for the analysis and merge. I will push the entropy for the rest of STM32 products soon.
Cheers
Armelle
